syntax = "proto2";

import "development/vndk/tools/header-checker/proto/abi_dump.proto";

package abi_diff;

message RecordTypeDiff {
  optional string type_stack = 1;
  optional abi_dump.RecordType old = 2;
  optional abi_dump.RecordType new = 3;
}

message EnumTypeDiff {
  optional string type_stack = 1;
  optional abi_dump.EnumType old = 2;
  optional abi_dump.EnumType new = 3;
}

message FunctionDeclDiff {
  // Template diffs are not required since in C++, they will show up in mangled
  // names and in C, templates are not supported.
  optional string type_stack = 1;
  optional abi_dump.FunctionDecl old = 2;
  optional abi_dump.FunctionDecl new = 3;
}

message GlobalVarDeclDiff {
  optional string type_stack = 1;
  optional abi_dump.GlobalVarDecl old = 2; // Old global var
  optional abi_dump.GlobalVarDecl new = 3; // New global var
}

message TranslationUnitDiff {
  // Library Name
  optional string lib_name = 1;
  optional string arch = 2;
  // Differing Elements.
  repeated RecordTypeDiff safe_record_type_diffs = 3;
  repeated RecordTypeDiff unsafe_record_type_diffs = 4;
  repeated FunctionDeclDiff unsafe_function_diffs = 5;
  repeated GlobalVarDeclDiff unsafe_global_var_diffs = 6;
  repeated FunctionDeclDiff safe_function_diffs = 7;
  repeated GlobalVarDeclDiff safe_global_var_diffs = 8;
  // Removed Elements.
  repeated abi_dump.FunctionDecl functions_removed = 9;
  repeated abi_dump.GlobalVarDecl global_vars_removed = 10;
  // Added Elements.
  repeated abi_dump.FunctionDecl functions_added = 11;
  repeated abi_dump.GlobalVarDecl global_vars_added = 12;
  // Compatiblity Status
  optional CompatibilityStatus compatibility_status = 13;
  repeated EnumTypeDiff safe_enum_type_diffs = 14;
  repeated EnumTypeDiff unsafe_enum_type_diffs = 15;
  repeated abi_dump.RecordType record_types_removed = 16;
  repeated abi_dump.EnumType enum_types_removed = 17;
}

// Not merged with TranslationUnitDiff to allow future extensions.
message ConciseDiffReportInformation {
  optional string lib_name = 1;
  optional string arch = 2;
  optional string diff_report_path = 3;
  optional CompatibilityStatus compatibility_status = 4;
}

message MergedTranslationUnitDiff {
  repeated ConciseDiffReportInformation diff_reports = 1;
}

enum CompatibilityStatus {
  COMPATIBLE = 0;
  EXTENSION = 1;
  INCOMPATIBLE = 4;
}
